connections/hole-punching.md

```diff
@@ -0,0 +1,257 @@
+
+- Both A and B discover [STUN] servers through some external mechanism. Given
+  that two hosts do not need to share the same [STUN] server to establish a
+  WebRTC connection, this does not require any coordination beyond the [STUN]
+  server discovery process.
+
+- B establishes a Websocket connections to one or more relay server and listens
+  for incoming connection requests via the [circuit relay v2
+  protocol][circuit-relay-v2]. Once B established a reservation at one or more
+  relay servers, B can advertise its relayed addresses (e.g.
+  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.
+
+- A discovers the relayed address of B through some external mechanism. Given
+  B's relayed address A can establish a TCP or QUIC connection to the relay
+  server of B. With the help of the [circuit relay v2
+  protocol][circuit-relay-v2] protocol A and B can exchange [SDP] messages
+  eventually allowing A B and to establish a WebRTC connection.
+
+### Private Browser (A) to Private Browser (B)
+
+- Both A and B discover [STUN] servers through some external mechanism. Given
+  that two hosts do not need to share the same [STUN] server to establish a
+  WebRTC connection, this does not require any coordination beyond the [STUN]
+  server discovery.
+
+- B establishes a Websocket connections to one or more relay servers and listens
+  for incoming connection requests via the [circuit relay v2
+  protocol][circuit-relay-v2]. Once B established a reservation at one or more
+  relay servers, B can advertise its relayed addresses (e.g.
+  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.
+
+- A discovers the relayed address of B through some external mechanism. Given
+  B's relayed address A can establish a Websocket connection to the relay server
+  of B. With the help of the [circuit relay v2 protocol][circuit-relay-v2]
+  protocol A and B can exchange [SDP] messages eventually allowing A B and to
+  establish a WebRTC connection.
+
+## FAQ
+
+- **Why not connect to a relay server via WebRTC instead of Websockets?**
+
+  WebRTC needs some mechanism to exchange [SDP] messages. Unless there is some
+  other means to exchange these [SDP] messages to a relay server, one has to use
+  Websockets.
+
+- **What to do when hole punching fails?**
+
+  Peers will be able to communicate via the established [circuit relay v2
+  protocol][circuit-relay-v2] connection. Though this connection is limited in
+  both transfered bytes and time. For now, upper layer protocols need to cope
+  with the fact that full connectivity can not be guaranteed.
+
+- **Why use both [AutoNAT] and [STUN], why not settle on one for both TCP / QUIC
+  and WebRTC based hole punching?**
+  
+  On browser platforms libp2p implementations do not control the local WebRTC
+  stack. This same stack can only operate with [STUN] and not [AutoNAT]. One
+  could use [STUN] instead of [AutoNAT] for TCP and QUIC hole punching though.
```

Comment:
So why do we have AutoNAT instead of STUN?  I clicked the link, but see it's a TODO to write this spec.

Reply:
I don't think I am the right person to reply here as I have not been involved in the design nor in one of the implementations. If I recall correctly one consideration was to prevent non-libp2p free-riding. @vyzo can you add more details here?

Reply:
in part yes, but it was more motivated for having pervasive support with no dedicated server.
Any public node can provide the AutoNAT service.

Note that AutoNAT is used only for detecting public reachability; the observed address inference from identify complements it.